rfd/0038-parquet-and-events.md

```diff
@@ -0,0 +1,139 @@
+
+Option 1 has multiple significant drawbacks which make it unsuitable for production use. The major issue is that all data is fundamentally tied to the only layer of security meaning that if a malicous party potentially acquired access hold to the key, all data would have to be reincrypted. This makes key rotation prohibitive. Another major issue is that of prolonged key reuse which contradicts industry best practices.
+
+Option 2 is doing something very similar to what the AWS Encryption SDK does. Ideally using that library would be ideal since it is a very nice encryption toolkit but unfortunately it isn't available for Go. The strategy involves using unique ephemeral data keys for each encrypted object (rotated when reincrypted/modified) that are encrypted and bundled with each encrypted object in a defined format. A central PEM encoded master keypair stored in a new resource is used to encrypt these per object keys.
+
+Option 2 here seems like a far more robust solution for production use and that is thus what we should employ.
+
+For facilitating key rotation the encrypted objects should be stored in a seperate file from it's data key bundle. This is due to modification restricts on objects imposed by services such as AWS S3.
+
+All encrypted objects should be suffixed by a `.enc` extension to communicate that they are encrypted. For simplicity no other data should be bundled in the same file.
+
+The file that stores the encrypted data key should be a file with the name `$objectName.key` stored next to the encrypted object.
+
+Each encrypted object has an accompanying file named `$objectName.key` that stores a JSON document of this format.
+
+```json
+{
+    // Contains a base64 encoded AES-CBC-HMACSHA256 data key encrypted with the master key.
+    "dataKey": "string",
+
+    // An integer starting at 1 one that communicates the schema used for encryption.
+    // May be changed in the future.
+    "version": "number",
+
+    // A hex encoded version of the SHA-256 hash of the master key.
+    // This is used to determine state during key rotation.
+    "masterKey": "string",
+}
+```
+
+This format allows a decrypting auth server to easily fetch and decrypt the data key using the master key. It can then fetch and decrypt the object itself.
+
+Key rotation in event of a security incident can be performed by fetching and rewriting every key file to reencrypt the data key.
+
+#### Encryption compatability
+
+To make Parquet encryption compatible with SIEM and data analysis tools when optionally enabled, data keys are optionally stored in AWS KMS and are tagged with an ID that is also included in the KEYID metadata of the Parquet files. This allows automatic key lookup and decryption.
+
+#### HSM Support
+
+HSM support will be implemented by loading master keys into the HSM for encrypting and decrypting data keys. They will also be used to securely generate those data keys and encrypt the data.
+
```

Comment:
In an HA setup, each auth server may be connected to a different HSM.

Consider a cluster with a single Auth server where the master key is held in an HSM. If you add a second auth server connected to a new HSM, I'm not sure how it will be able to decrypt the existing data keys (I'm assuming that would be necessary). Even adding new encrypted files in an HA HSM cluster would require some coordination. Might be worth thinking about this sooner rather than later if it needs to be supported

Reply:
Can we actually make this HSM-only design? Each auth server will encrypt it's portion of the stream using key stored in the stream and encrypted/decrypted using HSM's main key"?

Reply:
Another I suppose alternative option is to encrypt the master key in the backend once for every HSM key and only store it unencrypted in memory temporarily for managing data keys. This shared access makes adding new auths easy as long as you have an old auth since they can coordinate another encrypted master key easily via the backend.

Reply:
@xacrimon sounds to me like either of your solutions could work, but I don't unfortunately know the crypto or the requirements for this well enough to suggest one